k256/src/ecdsa.rs

```
@@ -173,6 +182,56 @@ pub type SigningKey = ecdsa_core::SigningKey<Secp256k1>;
 #[cfg(feature = "ecdsa")]
 pub type VerifyingKey = ecdsa_core::VerifyingKey<Secp256k1>;
 
+#[cfg(all(feature = "ecdsa", feature = "arithmetic"))]
+/// Extension trait for [`VerifyingKey`] for zkvm
+pub trait PowdrVerifyKey {
+    /// Recover a [`VerifyingKey`] from a pre-hashed message and signature.
+    fn powdr_recover_from_prehash(
+        prehash: &[u8],
+        signature: &Signature,
+        recovery_id: RecoveryId,
+    ) -> Result<VerifyingKey, Error>;
+}
+#[cfg(all(feature = "ecdsa", feature = "arithmetic"))]
+#[allow(non_snake_case)]
+impl PowdrVerifyKey for ecdsa_core::VerifyingKey<Secp256k1> {
+    fn powdr_recover_from_prehash(
+        prehash: &[u8],
+        signature: &Signature,
+        recovery_id: RecoveryId,
+    ) -> Result<Self, Error> {
+        let (r, s) = signature.split_scalars();
+        let z = <Scalar as Reduce<<Secp256k1 as Curve>::Uint>>::reduce_bytes(&bits2field::<
+            Secp256k1,
+        >(prehash)?);
+
+        let r_bytes = if recovery_id.is_x_reduced() {
+            let repr: FieldBytes = r.to_repr();
+            let r_bytes = <U as FieldBytesEncoding<Secp256k1>>::decode_field_bytes(&repr)
+                .checked_add(&ORDER)
+                .unwrap();
```

This is different from `VerifyingKey::recover_from_prehash` (panics instead of returning an error), I think it should be consistent?

---

Done